README.adoc - ID review - microprofile-telemetry-jaeger guide (#28)
OpenLiberty/guide-microprofile-telemetry-jaeger: README.adoc

:page-seo-description: A getting started tutorial and an example on how to enable distributed tracing in Java microservices to easily trace request flows that span multiple resources by using MicroProfile Telemetry and Jaeger distributed tracing system.​
:source-highlighter: prettify
:guide-author: Open Liberty

> = Enabling distributed tracing in microservices by using the OpenTelemetry observability framework 

Reviewer: I think we need a shorter title- also do we want to call out OT or Microprofile? Or Jaeger?
```suggestion
= Enabling distributed tracing in microservices with OpenTelemetry and Jaeger
```

Reviewer: @gkwan-ibm are there plans to deprecate the two existing guides that use MP OpenTracing? I believe the feature is now stabilized 

https://openliberty.io/guides/microprofile-opentracing.html
https://openliberty.io/guides/microprofile-opentracing-jaeger.html

[.hidden]

> NOTE: This repository contains the guide documentation source. To view the guide in published form, view it at the https://openliberty.io/guides/{projectid}.html[Open Liberty website].

Reviewer: I think this is ok as is

> Distributed tracing helps teams keep track of requests between microservices. MicroProfile Telemetry adopts OpenTelemetry tracing, allowing developers to observe requests across their distributed systems.

Reviewer: since the audience for the guide is developers, use "you" to address reader directly
```suggestion
Distributed tracing helps teams keep track of requests between microservices. MicroProfile Telemetry adopts OpenTelemetry tracing, so you can observe requests across your distributed systems.
```

// =================================================================================================
//  What you'll learn
// =================================================================================================

== What you'll learn

> The use of a microservices architecture can increase the difficulty of understanding how services depend on or affect other services. Therefore, It makes it harder to identify the source of latency or inaccuracies.

Reviewer: revise for clarity and extra words
```suggestion
The complexity of microservices architecture can make it more difficult to understand how services depend on or affect each other and to identify sources of latency or inaccuracies.
```

> One way to increase the observability of an application is by emitting traces. https://opentelemetry.io/[OpenTelemetry^] is a set of APIs, SDKs, tooling, and integrations designed to create and manage telemetry data such as traces, metrics, and logs. MicroProfile Telemetry adopts OpenTelemetry, so your applications can benefit from both manual and automatic traces.

Reviewer: ```suggestion
One way to increase the observability of an application is by emitting traces. https://opentelemetry.io/[OpenTelemetry^] is a set of APIs, SDKs, tooling, and integrations designed to create and manage telemetry data such as traces, metrics, and logs. MicroProfile Telemetry adopts OpenTelemetry so your applications can benefit from both manual and automatic traces.
```

> Traces represent requests, which can contain multiple operations or spans. Each span comprises a name, time-related data, log messages, and metadata that describes what occurred during a transaction. Spans are associated with a context, which identifies the request within which the span occurred. Developers can then follow a single request between services through a potentially complex distributed system. Exporters send the data that MicroProfile Telemetry 1.0 collects to Jaeger so you can visualize and monitor the generated spans.

Reviewer: ```suggestion
Traces represent requests, which can contain multiple operations or spans. Each span comprises a name, time-related data, log messages, and metadata that describe what occurred during a transaction. Spans are associated with a context, which identifies the request within which the span occurred. Developers can then follow a single request between services through a potentially complex distributed system. Exporters send the data that MicroProfile Telemetry 1.0 collects to Jaeger so you can visualize and monitor the generated spans.
```

use plural verb for multiple entities comprising a span

== Additional prerequisites 

> Before you begin, deploy the Jaeger all-in-one executable file to start the Jaeger tracing system. The Jaeger all-in-one executable file is configured for quick local testing. Jaeger 1.46 or later is recommended. You can find information about the Jaeger server, and the instructions for starting the all-in-one executable file in the https://www.jaegertracing.io/docs/latest/getting-started[Jaeger documentation].

Reviewer: I think this was better as it was

// cloud-hosted guide instructions:
ifdef::cloud-hosted[]

> Run the following Docker command to start the Jaeger server:

Reviewer: ```suggestion
Run the following `docker` command to start the Jaeger server:
```

----

ifndef::cloud-hosted[]
Make sure that your Jaeger server is running and navigate your browser to the http://localhost:9081/inventory/systems/localhost URL.

> When you visit this endpoint, make two GET HTTP requests, one to the `system` service and another to the `inventory` service. Both of these requests are configured to be traced, so a new trace is recorded in Jaeger. To view the traces, go to the http://localhost:16686 URL.

Reviewer: ```suggestion
When you visit this endpoint, you make two GET HTTP requests, one to the `system` service and another to the `inventory` service. Both of these requests are configured to be traced, so a new trace is recorded in Jaeger. To view the traces, go to the http://localhost:16686 URL.
```

curl -s http://localhost:9081/inventory/systems/localhost | jq
```

> When you visit this endpoint, make two GET HTTP requests, one to the ***system*** service and another to the ***inventory*** service. Both of these requests are configured to be traced, so a new trace is recorded in Jaeger. To view the traces, click the following button to visit the Jaeger service:

Reviewer: ```suggestion
When you visit this endpoint, you make two GET HTTP requests, one to the ***system*** service and another to the ***inventory*** service. Both of these requests are configured to be traced, so a new trace is recorded in Jaeger. To view the traces, click the following button to visit the Jaeger service:
```

::startApplication{port="16686" display="external" name="Visit Jaeger service" route="/"}
endif::[]

> You can view the traces for the `system` or `inventory` services under the **Search** tab. If you see only the **jaeger-query** option that is listed in the drop-down, wait a little longer and refresh the page to see the application services.

Reviewer: ```suggestion
You can view the traces for the `system` or `inventory` services under the **Search** tab. If you see only the **jaeger-query** option in the drop-down menu, wait a little longer and refresh the page to see the application services.
```

Select the services in the **Select A Service** menu and click the **Find Traces** button at the end of the section. You will see the following result:

image::inventory_service_spans.png[Get traces for the inventory service,align="center"]

> The trace has five spans, four from the `inventory` service and one from the `system` service. Click the trace to view its details. Under **Service & Operation**, you see the spans in this trace. You can inspect each span by clicking it to reveal more detailed information, such as the time at which a request was received and the time at which a response was sent back.

Reviewer: ```suggestion
The trace has five spans, four from the `inventory` service and one from the `system` service. Click the trace to view its details. Under **Service & Operation**, you see the spans in this trace. You can inspect each span by clicking it to reveal more detailed information, such as the times that a request was received and a response was sent.
```

The [hotspot=mpTelemetry file=1]`mpTelemetry` feature is now enabled in the `server.xml` of the `inventory` service.

> By default, MicroProfile Telemetry tracing is `off`. To enable any tracing aspects, specify the `otel` properties in the MicroProfile configuration file. 

Reviewer: ```suggestion
By default, MicroProfile Telemetry tracing is off. To enable any tracing aspects, specify the `otel` properties in the MicroProfile configuration file. 
```

"off" is not a code entity

> In order to access the Tracer, the [hotspot=inject file=0]`@Inject` annotation from the Contexts and Dependency Injections API injects the Tracer into a bean. 

Reviewer: ```suggestion
To access the Tracer, the [hotspot=inject file=0]`@Inject` annotation from the Contexts and Dependency Injections API injects the Tracer into a bean. 
```

> Before the [hotspot=manager hotspot=getSystem file=0]`InventoryManager` calls the `system` service, it creates and starts a span called the [hotspot=getPropertiesSpan file=0]`GettingProperties` span by using the `spanBuilder()` and `startSpan()` Tracer APIs.

Reviewer: i think this was better before- dont need to say span twice

> When starting a span, you also need to end it by calling [hotspot=end file=0]`end()` on the span. If you don't end a span, it won't be recorded at all and won't show up in Jaeger. This code ensures that [hotspot=end file=0]`end()` is always called by including it in a `finally` block.

Reviewer: ```suggestion
When you start a span, you must also end it by calling [hotspot=end file=0]`end()` on the span. If you don't end a span, it won't be recorded at all and won't show up in Jaeger. This code ensures that [hotspot=end file=0]`end()` is always called by including it in a `finally` block.
```

> After starting the span, you make it current with the [hotspot=scope file=0]`makeCurrent()` call. Making a span current means that any new spans created in the same thread (either automatically by open liberty, or manually by calling the API) will use this span as their parent span.

Reviewer: ```suggestion
After you start the span, make it current with the [hotspot=scope file=0]`makeCurrent()` call. Making a span current means that any new spans created in the same thread, either automatically by open liberty or manually by calling the API, will use this span as their parent span.
```

> The `makeCurrent()` call returns a `Scope`. It is important to always close the `Scope`, which stops the span from being current and makes the previous span current again. The recommended way to do this is to use a [hotspot=try file=0]`try-with-resources` block, which makes sure that the `Scope` is automatically closed at the end of the block.

Reviewer: ```suggestion
The `makeCurrent()` call returns a `Scope`. Make sure to always close the `Scope`, which stops the span from being current and makes the previous span current again. Use a [hotspot=try file=0]`try-with-resources` block, which automatically closes the `Scope` at the end of the block.
```

// =================================================================================================
== Testing the application 

> Manually verify the traces by inspecting them on the Jaeger server. You will find some tests included to assess the basic functionality of the services. If any of the tests fail, it might indicate a potential bug in the code.

Reviewer: ```suggestion
Manually verify the traces by inspecting them on the Jaeger server. You will find some tests included to test the basic functionality of the services. If any of the tests fail, you might have introduced a bug into the code.
```
